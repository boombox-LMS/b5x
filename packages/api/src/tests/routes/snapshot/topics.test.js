const supertest = require("supertest");

describe("Topics routes should match expectations", () => {
  let app;
  let cookie = "";
  let testTopicUri = global.SMOKE_TEST_TOPIC_URI;
  const apiPrefix = global.__SUPERTEST_API_PREFIX__;

  beforeAll(async () => {
    // Add a seeded test database to the app, available as app.get('db')
    app = await global.buildAppWithSeededTestDatabase({
      users: [{ persona: "completed" }],
    });

    // Log the user in
    await supertest(app)
      .post(apiPrefix + "users.logIn")
      .send({ email: "completed-1@test.com" })
      .then((res) => {
        // Save the cookie to send with later requests
        cookie = global.getCookie(res) || cookie;
      });
  });

  // Close the database connection, which allows Jest to exit gracefully
  afterAll(async () => {
    await app.get("db").destroy();
  });

  // topics.catalog -------------------------------------------------

  describe("topics.catalog matches the snapshot", () => {
    let responseBody;

    test("topics.catalog returns a response", async () => {
      await supertest(app)
        .get(apiPrefix + "topics.catalog")
        .set("Cookie", cookie)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("topics.catalog matches the snapshot", () => {
      expect(responseBody).toMatchSnapshot();
    });
  });

  // topics.info ----------------------------------------------------

  describe("topics.info matches expectations", () => {
    let responseBody;

    test("topics.info returns a response", async () => {
      await supertest(app)
        .get(apiPrefix + `topics.info?uri=${testTopicUri}`)
        .set("Cookie", cookie)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("topics.info matches the snapshot on file", async () => {
      expect(responseBody).toMatchSnapshot();
    });
  });

  // topics.contents ------------------------------------------------

  describe("topics.contents matches the snapshot", () => {
    let responseBody;

    test("topics.contents returns a response", async () => {
      await supertest(app)
        .get(apiPrefix + `topics.contents?uri=${testTopicUri}`)
        .set("Cookie", cookie)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("topics.contents matches the snapshot on file", () => {
      expect(responseBody).toMatchSnapshot();
    });
  });

  // topics.enrollment + topics.verifyCompletion --------------------

  describe("topics.enrollment and topics.verifyCompletion match the snapshot", () => {
    let enrollmentResponseBody;
    let completionResponseBody;

    test("topics.enrollment and topics.verifyCompletion return a response", async () => {
      // trigger enrollment creation
      await supertest(app)
        .get(apiPrefix + `topics.enrollment?uri=${testTopicUri}`)
        .set("Cookie", cookie)
        .then((res) => {
          enrollmentResponseBody = res.body;
        });

      // verify topic completion
      await supertest(app)
        .post(apiPrefix + "topics.verifyCompletion")
        .set("Cookie", cookie)
        .send({ topicUri: testTopicUri })
        .then((res) => {
          completionResponseBody = res.body;
        });
    });

    test("topics.enrollment matches the snapshot", () => {
      expect(enrollmentResponseBody).toMatchSnapshot();
    });

    test("topics.verifyCompletion matches the snapshot", () => {
      expect(completionResponseBody).toMatchSnapshot();
    });
  });
});
