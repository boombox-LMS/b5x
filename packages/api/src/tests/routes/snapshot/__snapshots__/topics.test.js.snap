// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topics routes should match expectations topics.catalog matches the snapshot topics.catalog matches the snapshot 1`] = `
{
  "filters": {
    "completionStatus": {
      "completed": false,
      "in progress": true,
      "not started": true,
    },
    "priorityLevel": {
      "assigned": true,
      "available": false,
      "recommended": true,
    },
  },
  "topics": [
    {
      "completionStatus": "completed",
      "contentMode": "documentation",
      "coverImageUrl": "http://localhost:8080/topic-assets/boombox-feature-utopia-vseed/images/content-utopia.jpg",
      "currentDocumentUri": "boombox-feature-utopia-vseed_achievement-badge",
      "firstDocumentUri": "boombox-feature-utopia-vseed_achievement-badge",
      "prerequisites": [],
      "priorityLevel": "recommended",
      "progressPercentage": 0,
      "slug": "boombox-feature-utopia",
      "subtitle": "The full list of content components.",
      "title": "Boombox Component Catalog",
      "unmetPrerequisites": [],
      "uri": "boombox-feature-utopia-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-5-vseed/images/building-to-scale.jpg",
      "currentDocumentUri": "filler-topic-5-vseed_doc-1",
      "firstDocumentUri": "filler-topic-5-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-5",
      "subtitle": "A primer on microservices and other enterprise architecture",
      "title": "Building Apps to Scale",
      "unmetPrerequisites": [],
      "uri": "filler-topic-5-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/code-of-conduct-vseed/images/code-of-conduct.jpg",
      "currentDocumentUri": "code-of-conduct-vseed_doc-1",
      "firstDocumentUri": "code-of-conduct-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "assigned",
      "progressPercentage": 0,
      "slug": "code-of-conduct",
      "subtitle": "A universal Code of Conduct that can be required by other topics as a prerequisite",
      "title": "Code of Conduct",
      "unmetPrerequisites": [],
      "uri": "code-of-conduct-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/data-driven-decisionmaking-vseed/images/data-at-cloudco.jpg",
      "currentDocumentUri": "data-driven-decisionmaking-vseed_customization-page",
      "firstDocumentUri": "data-driven-decisionmaking-vseed_customization-page",
      "prerequisites": [
        "code-of-conduct",
      ],
      "priorityLevel": "recommended",
      "progressPercentage": 0,
      "slug": "data-driven-decisionmaking",
      "subtitle": "An introduction to SQL, data visualization, and experimentation.",
      "title": "Data-Driven Decisionmaking",
      "unmetPrerequisites": [],
      "uri": "data-driven-decisionmaking-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-4-vseed/images/heroku.jpeg",
      "currentDocumentUri": "filler-topic-4-vseed_doc-1",
      "firstDocumentUri": "filler-topic-4-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-4",
      "subtitle": "Using command-line tools to ship an app",
      "title": "Deploying on Heroku",
      "unmetPrerequisites": [],
      "uri": "filler-topic-4-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-11-vseed/images/experimentation.jpg",
      "currentDocumentUri": "filler-topic-11-vseed_doc-1",
      "firstDocumentUri": "filler-topic-11-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-11",
      "subtitle": "Building, deploying, and analyzing experiments.",
      "title": "Experimentation",
      "unmetPrerequisites": [],
      "uri": "filler-topic-11-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-9-vseed/images/incident-management.jpg",
      "currentDocumentUri": "filler-topic-9-vseed_doc-1",
      "firstDocumentUri": "filler-topic-9-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-9",
      "subtitle": "Our worst incidents (and what we learned from them).",
      "title": "Incident Case Studies",
      "unmetPrerequisites": [],
      "uri": "filler-topic-9-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/intro-to-boombox-lms-vseed/images/boombox.jpg",
      "currentDocumentUri": "intro-to-boombox-lms-vseed_skills-and-customization",
      "firstDocumentUri": "intro-to-boombox-lms-vseed_skills-and-customization",
      "prerequisites": [],
      "priorityLevel": "assigned",
      "progressPercentage": 0,
      "slug": "intro-to-boombox-lms",
      "subtitle": "What is it? Why is it? How does it work? Your questions, answered.",
      "title": "Intro to Boombox LMS",
      "unmetPrerequisites": [],
      "uri": "intro-to-boombox-lms-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-6-vseed/images/kibana.jpeg",
      "currentDocumentUri": "filler-topic-6-vseed_doc-1",
      "firstDocumentUri": "filler-topic-6-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-6",
      "subtitle": "Common queries and visualizations",
      "title": "Intro to Kibana",
      "unmetPrerequisites": [],
      "uri": "filler-topic-6-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-1-vseed/images/sql.jpg",
      "currentDocumentUri": "filler-topic-1-vseed_doc-1",
      "firstDocumentUri": "filler-topic-1-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-1",
      "subtitle": "A beginner's guide to querying data in Mode",
      "title": "Intro to SQL",
      "unmetPrerequisites": [],
      "uri": "filler-topic-1-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-8-vseed/images/robot-friend.jpg",
      "currentDocumentUri": "filler-topic-8-vseed_doc-1",
      "firstDocumentUri": "filler-topic-8-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-8",
      "subtitle": "A proprietary challenge set for graduates of Intro to ML",
      "title": "ML Challenges",
      "unmetPrerequisites": [],
      "uri": "filler-topic-8-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-10-vseed/images/observability.jpg",
      "currentDocumentUri": "filler-topic-10-vseed_doc-1",
      "firstDocumentUri": "filler-topic-10-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-10",
      "subtitle": "A step-by-step guide to designing a new feature with observability in mind.",
      "title": "Observability Planning",
      "unmetPrerequisites": [],
      "uri": "filler-topic-10-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-12-vseed/images/sdlc.jpg",
      "currentDocumentUri": "filler-topic-12-vseed_doc-1",
      "firstDocumentUri": "filler-topic-12-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-12",
      "subtitle": "Our processes and policies for shipping high-quality features.",
      "title": "Our SDLC",
      "unmetPrerequisites": [],
      "uri": "filler-topic-12-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-13-vseed/images/python-best-practices.jpg",
      "currentDocumentUri": "filler-topic-13-vseed_doc-1",
      "firstDocumentUri": "filler-topic-13-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-13",
      "subtitle": "Mocks, fixtures, and more",
      "title": "Python Best Practices",
      "unmetPrerequisites": [],
      "uri": "filler-topic-13-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-7-vseed/images/security.jpg",
      "currentDocumentUri": "filler-topic-7-vseed_doc-1",
      "firstDocumentUri": "filler-topic-7-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-7",
      "subtitle": "Patterns for avoiding the most common attacks",
      "title": "Security Best Practices",
      "unmetPrerequisites": [],
      "uri": "filler-topic-7-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/smoke-testing-vseed/images/smoke-testing.jpg",
      "currentDocumentUri": "smoke-testing-vseed_setup",
      "firstDocumentUri": "smoke-testing-vseed_setup",
      "prerequisites": [],
      "priorityLevel": "assigned",
      "progressPercentage": 0,
      "slug": "smoke-testing",
      "subtitle": "Verify that this topic still works before merging to main.",
      "title": "Smoke Testing",
      "unmetPrerequisites": [],
      "uri": "smoke-testing-vseed",
      "version": "seed",
    },
    {
      "completionStatus": "completed",
      "contentMode": "tutorial",
      "coverImageUrl": "http://localhost:8080/topic-assets/filler-topic-2-vseed/images/redux.png",
      "currentDocumentUri": "filler-topic-2-vseed_doc-1",
      "firstDocumentUri": "filler-topic-2-vseed_doc-1",
      "prerequisites": [],
      "priorityLevel": "available",
      "progressPercentage": 0,
      "slug": "filler-topic-2",
      "subtitle": "A guide to managing React state with Redux Toolkit",
      "title": "Using Redux Toolkit",
      "unmetPrerequisites": [],
      "uri": "filler-topic-2-vseed",
      "version": "seed",
    },
  ],
}
`;

exports[`Topics routes should match expectations topics.contents matches the snapshot topics.contents matches the snapshot on file 1`] = `
{
  "config": {
    "contentMode": "tutorial",
    "groupAccessLevels": {
      "assigned": [
        "all",
      ],
    },
    "prerequisites": [],
    "slug": "smoke-testing",
  },
  "coverImageUrl": "http://localhost:8080/topic-assets/smoke-testing-vseed/images/smoke-testing.jpg",
  "documents": [
    {
      "title": "Setup",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_setup",
    },
    {
      "title": "Progress bar",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_progress-bar",
    },
    {
      "title": "Major UI areas",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_ui-areas",
    },
    {
      "title": "Header",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_header",
    },
    {
      "title": "User state",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_persistence",
    },
    {
      "title": "Conditional content display",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_conditional-display",
    },
    {
      "title": "ONE",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_one",
    },
    {
      "title": "Logging in and out",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_login-logout",
    },
    {
      "title": "Dashboard",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_dashboard-stats",
    },
    {
      "title": "Publishing a topic",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_publishing",
    },
    {
      "title": "Topic completion",
      "topicUri": "smoke-testing-vseed",
      "uri": "smoke-testing-vseed_topic-completion",
    },
  ],
  "prerequisites": [],
  "slug": "smoke-testing",
  "subtitle": "Verify that this topic still works before merging to main.",
  "title": "Smoke Testing",
  "uri": "smoke-testing-vseed",
}
`;

exports[`Topics routes should match expectations topics.enrollment and topics.verifyCompletion match the snapshot topics.enrollment matches the snapshot 1`] = `
{
  "currentDocumentUri": "smoke-testing-vseed_setup",
  "documentStatus": {
    "smoke-testing-vseed_conditional-display": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_one",
    },
    "smoke-testing-vseed_dashboard-stats": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_publishing",
    },
    "smoke-testing-vseed_header": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_persistence",
    },
    "smoke-testing-vseed_login-logout": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_dashboard-stats",
    },
    "smoke-testing-vseed_one": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_login-logout",
    },
    "smoke-testing-vseed_persistence": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_conditional-display",
    },
    "smoke-testing-vseed_progress-bar": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_ui-areas",
    },
    "smoke-testing-vseed_publishing": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_topic-completion",
    },
    "smoke-testing-vseed_setup": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_progress-bar",
    },
    "smoke-testing-vseed_topic-completion": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": null,
    },
    "smoke-testing-vseed_ui-areas": {
      "isCompleted": true,
      "isLocked": false,
      "isVisible": true,
      "nextVisibleDocumentUri": "smoke-testing-vseed_header",
    },
  },
  "id": 20,
  "progressPercentage": 100,
  "topicIsCompleted": true,
  "topicUri": "smoke-testing-vseed",
  "userId": 2,
}
`;

exports[`Topics routes should match expectations topics.enrollment and topics.verifyCompletion match the snapshot topics.verifyCompletion matches the snapshot 1`] = `
{
  "topicIsCompleted": true,
  "topicUri": "smoke-testing-vseed",
}
`;

exports[`Topics routes should match expectations topics.info matches expectations topics.info matches the snapshot on file 1`] = `
{
  "config": {
    "contentMode": "tutorial",
    "groupAccessLevels": {
      "assigned": [
        "all",
      ],
    },
    "prerequisites": [],
    "slug": "smoke-testing",
  },
  "coverImageUrl": "http://localhost:8080/topic-assets/smoke-testing-vseed/images/smoke-testing.jpg",
  "enrollmentData": {
    "currentDocumentUri": "smoke-testing-vseed_setup",
    "topicVersion": "seed",
  },
  "firstDocumentUri": "smoke-testing-vseed_setup",
  "prerequisites": [],
  "slug": "smoke-testing",
  "subtitle": "Verify that this topic still works before merging to main.",
  "title": "Smoke Testing",
  "unmetPrerequisites": [],
  "uri": "smoke-testing-vseed",
  "version": "seed",
}
`;
